Chapter02/Exercise2.03/app/src/main/java/com/example/saveandrestore/MainActivity.kt

```diff
@@ -4,23 +4,49 @@ import android.content.Context
 import android.os.Bundle
 import android.util.Log
 import android.view.inputmethod.InputMethodManager
+import android.widget.Button
+import android.widget.EditText
+import android.widget.TextView
 import android.widget.Toast
 import androidx.appcompat.app.AppCompatActivity
-import kotlinx.android.synthetic.main.activity_main.*
 import java.util.*
 
 class MainActivity : AppCompatActivity() {
 
+    private val discountButton by lazy {
+        findViewById<Button>(R.id.discount_button)
+    }
+
+    private val firstName by lazy {
+        findViewById<EditText>(R.id.first_name)
+    }
+
+    private val lastName by lazy {
+        findViewById<EditText>(R.id.last_name)
+    }
+
+    private val email by lazy {
+        findViewById<EditText>(R.id.email)
+    }
+
+    private val discountCodeConfirmation by lazy {
+        findViewById<TextView>(R.id.discount_code_confirmation)
+    }
+
+    private val discountCode by lazy {
+        findViewById<TextView>(R.id.discount_code)
+    }
+
     override fun onCreate(savedInstanceState: Bundle?) {
         super.onCreate(savedInstanceState)
         setContentView(R.layout.activity_main)
         Log.d(TAG, "onCreate")
 
         // here we handle the Button onClick event
-        discount_button.setOnClickListener {
+        discountButton.setOnClickListener {
 
-            val firstName = first_name.text.toString().trim()
-            val lastName = last_name.text.toString().trim()
+            val firstName = firstName.text.toString().trim()
```

Comment:
same comment here for changing the name of these local variables to not make them the same as the field.